src/engine/input/key_mapping/InputActionItem.cs

```diff
@@ -0,0 +1,141 @@
+using System;
+using System.Collections.ObjectModel;
+using System.Collections.Specialized;
+using Godot;
+using Newtonsoft.Json;
+
+/// <summary>
+///   Defines one specific input.
+///   Each InputActionItem has <see cref="InputEventItem">InputEventItems</see> associated with it.
+/// </summary>
+[JsonConverter(typeof(InputActionItemConverter))]
```

Comment:
Could you explain where, and why you are using serialization on these GUI nodes?
This feels wrong to me. The GUI state should be easy enough to construct from the key data, right?

Reply:
The loading begins in this setter: https://github.com/Revolutionary-Games/Thrive/blob/9ebde8868546a8b9e8c51ed6a0848ba3f03a3c6d/src/engine/input/key_mapping/InputGroupList.cs#L51-L72
The setter is called from the InitFromData method: https://github.com/Revolutionary-Games/Thrive/blob/9ebde8868546a8b9e8c51ed6a0848ba3f03a3c6d/src/engine/input/key_mapping/InputGroupList.cs#L184-L198

It deserializes the input_options.json and stores it in a InputGroupItem List.
InputGroupItem has an JsonConverter associated with it to produce a godot ready instance of the InputGroupItem class.
InputGroupItem has those serializable properties:
- GroupName
- EnvironmentId
- Actions
Actions is an InputActionItem list and therefore triggers the InputActionItemConverter which produces a godot ready instance InputActionItem.
InputActionItem has those serializable properties:
- InputName
- DisplayName
- Inputs
Inputs is and ObservableCollection of InputEventItems.
Inputs and AssociatedGroup are set in the InputGroupItemConverter.
The InputEvent does not have an associated Converter, but gets created in the InputGroupItemConverter.

Somehow I have to load input_options.json and convert the data to godot objects. The simplest way I could imagine were JsonConverters

Reply:
I get how it works now. So here's a suggestion how to do it right:
Make new classes to represent the data in the settings JSON, then make a method for creating the necessary GUI controls for the data.
This can be further split up so that the child objects of the top level GUI can also have a method "ApplyFrom(object loaded from json)" so each level of objects delegates the construction of the GUI state to their children as well.

Reply:
hmm, I have notices that having those changes lead to a big issue: I need to update the InputEvents and cannot create them one time.
I'm kinda having problems with created a class structure...

Reply:
Here's a very rough sketch based on the JSON:

public class NamedInputAction{
public string InputName;
public string Name;
}

public class InputGroup{
public List<string> EnvironmentId;
public string GroupName;
public List<NamedInputAction> Actions;
}

Then you deserialize List<InputGroup>.
That should work.

If you want to get fancy you could make EnvironmentId into an enum so the json parse would validate that there are valid values in there.

Reply:
wouldn't it make more sense to make those structs?

Reply:
No. structs are value types in C# meaning that they aren't reference counted and the entire objects are copied around whenever they are passed to functions etc. I wouldn't use a struct unless there were very specific requirements on the code in C#.

Reply:
yeah, but structs also go to the stack instead of the heap.
But you're right, using classes I can mess up less :P